Quadriga/src/main/java/edu/asu/spring/quadriga/web/workspace/AddTextController.java

```diff
@@ -0,0 +1,123 @@
+    @RequestMapping(value = "/auth/workbench/workspace/{projectid}/{workspaceid}/addtext", method = RequestMethod.GET)
+    public ModelAndView addTextFileForm(@PathVariable("workspaceid") String workspaceid,
+            @PathVariable("projectid") String projid) throws QuadrigaStorageException, QuadrigaAccessException {
+
+        ModelAndView model = new ModelAndView("auth/workbench/workspace/addtext");
+        model.getModelMap().put("textfile", textFileFactory.createTextFileObject());
+        model.getModelMap().put("workspaceId", workspaceid);
+        model.getModelMap().put("myProjectId", projid);
+        model.getModelMap().put("success", "0");
+        return model;
+    }
+
+    /**
+     * @param resp
+     *            Generic HTTP Response Object
+     * @param txtFile
+     *            Web backing domain object to store the Text Details
+     * @param result
+     *            Generic BingingResults object to hold validation errors
+     * @param workspaceid
+     *            Path Variable to retrieve the associated Workspace id
+     * @return
+     * @throws QuadrigaStorageException
+     * @throws QuadrigaAccessException
+     * @throws IOException
+     */
+    @AccessPolicies({
+            @ElementAccessPolicy(type = CheckedElementType.WORKSPACE, paramIndex = 4, userRole = {
+                    RoleNames.ROLE_WORKSPACE_COLLABORATOR_ADMIN, RoleNames.ROLE_WORKSPACE_COLLABORATOR_ADMIN }),
+            @ElementAccessPolicy(type = CheckedElementType.PROJECT, paramIndex = 5, userRole = {
+                    RoleNames.ROLE_COLLABORATOR_ADMIN, RoleNames.ROLE_PROJ_COLLABORATOR_ADMIN,
+                    RoleNames.ROLE_PROJ_COLLABORATOR_CONTRIBUTOR }) })
+    @RequestMapping(value = "/auth/workbench/workspace/{projectid}/{workspaceid}/addtext", method = RequestMethod.POST)
+    public ModelAndView saveTextFileForm(HttpServletResponse resp,
+            @Validated @ModelAttribute("textfile") TextFile txtFile, BindingResult result,
+            @PathVariable("workspaceid") String workspaceid, @PathVariable("projectid") String projid)
+                    throws QuadrigaStorageException, QuadrigaAccessException, IOException {
+
+        ModelAndView model = new ModelAndView("auth/workbench/workspace/addtext");
+        if (result.hasErrors()) {
+            model.getModelMap().put("textfile", txtFile);
+            model.getModelMap().put("workspaceId", workspaceid);
+            model.getModelMap().put("myProjectId", projid);
+            model.getModelMap().put("success", "0");
+        } else {
+            model.getModelMap().put("textfile", textFileFactory.createTextFileObject());
+            model.getModelMap().put("workspaceId", workspaceid);
+            model.getModelMap().put("myProjectId", projid);
+            txtFile.setWorkspaceId(workspaceid);
+            txtFile.setProjectId(projid);
+            txtFileService.saveTextFile(txtFile);
+            model.getModelMap().put("success", "1");
```

Comment:
after adding a text file, the user should be redirected to the workspace details page and see a success flash message